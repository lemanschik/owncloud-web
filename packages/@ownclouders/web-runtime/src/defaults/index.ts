import merge from 'lodash-es/merge'
import App from '../App.vue'
import missingOrInvalidConfigPage from '../pages/missingOrInvalidConfig.vue'
import { coreTranslations, clientTranslations, pkgTranslations, odsTranslations } from './json'

// fontawesome-free attributions console message
import '@fortawesome/fontawesome-free/attribution'

export { default as DesignSystem } from 'design-system'

export const pages = { success: App, failure: missingOrInvalidConfigPage }
export const translations = merge(
  {},
  coreTranslations,
  clientTranslations,
  pkgTranslations,
  odsTranslations
)
export const supportedLanguages = {
  ar: 'اَلْعَرَبِيَّةُ',
  bg: 'български',
  en: 'English',
  de: 'Deutsch',
  es: 'Español',
  cs: 'Czech',
  fr: 'Français',
  gl: 'Galego',
  he: 'עִבְרִית',
  it: 'Italiano',
  pl: 'Polski',
  ru: 'русский язык',
  sk: 'Slovenčina',
  sq: 'Shqipja',
  tr: 'Türkçe'
}
