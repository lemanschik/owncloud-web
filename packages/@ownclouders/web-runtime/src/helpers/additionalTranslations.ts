// just a dummy function to trick gettext tools
function $gettext(msg: string): string {
  return msg
}

export const additionalTranslations = {
  admin: $gettext('Admin'),
  spaceAdmin: $gettext('Space Admin'),
  user: $gettext('User'),
  guest: $gettext('Guest'),
  activities: $gettext('Activities'),
  noActivities: $gettext('No activities'),
  virusDetectedActivity: $gettext(
    'Virus "%{description}" detected. Please contact your administrator for more information.'
  ),
  virusScan: $gettext('Scan for viruses')
}
