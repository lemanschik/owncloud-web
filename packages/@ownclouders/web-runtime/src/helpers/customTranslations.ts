import { ConfigurationManager } from 'web-pkg'
import { v4 as uuidV4 } from 'uuid'
import merge from 'lodash-es/merge'

export const loadCustomTranslations = async ({
  configurationManager
}: {
  configurationManager: ConfigurationManager
}): Promise<unknown> => {
  const customTranslations = {}
  for (const customTranslation of configurationManager.customTranslations) {
    const customTranslationResponse = await fetch(customTranslation.url, {
      headers: { 'X-Request-ID': uuidV4() }
    })
    if (customTranslationResponse.status !== 200) {
      console.error(
        `translation file ${customTranslation} could not be loaded. HTTP status-code ${customTranslationResponse.status}`
      )
      continue
    }
    try {
      const customTranslationJSON = await customTranslationResponse.json()
      merge(customTranslations, customTranslationJSON)
    } catch (e) {
      console.error(`translation file ${customTranslation} could not be parsed. ${e}`)
    }
  }
  return customTranslations
}
