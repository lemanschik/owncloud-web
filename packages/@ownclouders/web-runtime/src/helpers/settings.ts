export interface SettingsValue {
  identifier: {
    bundle: string
    extension: string
    setting: string
  }
  value: {
    accountUuid: string
    bundleId: string
    id: string
    resource: {
      type: string
    }
    settingId: string
    boolValue?: boolean
    listValue?: {
      values: {
        stringValue: string
      }[]
    }
  }
}

export interface SettingsBundle {
  displayName: string
  extension: string
  id: string
  name: string
  resource: {
    type: string
  }
  settings: {
    description: string
    displayName: string
    id: string
    name: string
    resource: {
      type: string
    }
    singleChoiceValue?: {
      options: Record<string, any>[]
    }
    boolValue?: Record<string, any>
  }[]
  type: string
}

export interface LanguageOption {
  label: string
  value: string
}
