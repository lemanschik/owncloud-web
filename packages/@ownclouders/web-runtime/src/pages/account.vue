<template>
  <app-loading-spinner v-if="isLoading" />
  <main v-else id="account" class="oc-height-1-1 oc-m">
    <div class="oc-flex oc-flex-between oc-flex-bottom oc-width-1-1 oc-border-b oc-py">
      <h1 id="account-page-title" class="oc-page-title oc-m-rm">{{ pageTitle }}</h1>
      <div>
        <edit-password-modal
          v-if="editPasswordModalOpen"
          @cancel="closeEditPasswordModal"
          @confirm="editPassword"
        ></edit-password-modal>
        <oc-button
          v-if="isChangePasswordEnabled"
          variation="primary"
          data-testid="account-page-edit-url-btn"
          @click="showEditPasswordModal"
        >
          <oc-icon name="lock" />
          <span v-text="$gettext('Change Password')" />
        </oc-button>
        <oc-button
          v-if="accountEditLink"
          variation="primary"
          type="a"
          :href="accountEditLink.href"
          target="_blank"
          data-testid="account-page-edit-url-btn"
        >
          <oc-icon name="edit" />
          <span v-text="$gettext('Edit')" />
        </oc-button>
      </div>
    </div>
    <h2 class="oc-text-bold oc-mb" v-text="$gettext('Account Information')" />
    <dl class="account-page-info oc-flex oc-flex-wrap">
      <div class="account-page-info-username oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Username')" />
        <dd>
          {{ user.username || user.id }}
        </dd>
      </div>
      <div v-if="user.username && user.id" class="account-page-info-userid">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('User ID')" />
        <dd>
          {{ user.id }}
        </dd>
      </div>
      <div class="account-page-info-displayname oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Display name')" />
        <dd>
          {{ user.displayname }}
        </dd>
      </div>
      <div class="account-page-info-email oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Email')" />
        <dd>
          <template v-if="user.email">{{ user.email }}</template>
          <span v-else v-text="$gettext('No email has been set up')" />
        </dd>
      </div>
      <div class="account-page-info-groups oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Group memberships')" />
        <dd data-testid="group-names">
          <span v-if="groupNames">{{ groupNames }}</span>
          <span
            v-else
            data-testid="group-names-empty"
            v-text="$gettext('You are not part of any group')"
          />
        </dd>
      </div>
      <div
        v-if="isSettingsServiceSupported"
        class="account-page-info-language oc-mb oc-width-1-2@s"
      >
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Language')" />
        <dd data-testid="language">
          <oc-select
            v-if="languageOptions"
            :placeholder="$gettext('Please choose...')"
            :model-value="selectedLanguageValue"
            :clearable="false"
            :options="languageOptions"
            @update:model-value="updateSelectedLanguage"
          />
        </dd>
      </div>
      <div v-if="logoutUrl" class="account-page-logout-all-devices oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Logout from active devices')" />
        <dd data-testid="logout">
          <oc-button
            appearance="raw"
            type="a"
            :href="logoutUrl"
            target="_blank"
            data-testid="account-page-logout-url-btn"
          >
            <span v-text="$gettext('Show devices')" />
          </oc-button>
        </dd>
      </div>
      <div v-if="showGdprExport" class="account-page-gdpr-export oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('GDPR export')" />
        <dd data-testid="gdpr-export">
          <gdpr-export />
        </dd>
      </div>
      <div v-if="isSettingsServiceSupported" class="account-page-notification oc-mb oc-width-1-2@s">
        <dt class="oc-text-normal oc-text-muted" v-text="$gettext('Notifications')" />
        <dd data-testid="notification-mails">
          <oc-checkbox
            :model-value="disableEmailNotificationsValue"
            size="large"
            :label="$gettext('Receive notification mails')"
            data-testid="account-page-notification-mails-checkbox"
            @update:model-value="updateDisableEmailNotifications"
          />
        </dd>
      </div>
    </dl>
  </main>
</template>

<script lang="ts">
import { mapActions } from 'vuex'
import EditPasswordModal from '../components/EditPasswordModal.vue'
import { SettingsBundle, LanguageOption, SettingsValue } from '../helpers/settings'
import { computed, defineComponent, onMounted, unref, ref } from 'vue'
import {
  useCapabilityGraphPersonalDataExport,
  useCapabilitySpacesEnabled,
  useClientService,
  useStore
} from 'web-pkg/src/composables'
import { useTask } from 'vue-concurrency'
import { useGettext } from 'vue3-gettext'
import { setCurrentLanguage } from 'web-runtime/src/helpers/language'
import GdprExport from 'web-runtime/src/components/Account/GdprExport.vue'
import { useConfigurationManager } from 'web-pkg/src/composables/configuration'
import { isPersonalSpaceResource } from 'web-client/src/helpers'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'

export default defineComponent({
  name: 'AccountPage',
  components: {
    AppLoadingSpinner,
    EditPasswordModal,
    GdprExport
  },
  setup() {
    const store = useStore()
    const language = useGettext()
    const { $gettext } = language
    const clientService = useClientService()
    const configurationManager = useConfigurationManager()
    const valuesList = ref<SettingsValue[]>()
    const accountBundle = ref<SettingsBundle>()
    const selectedLanguageValue = ref<LanguageOption>()
    const disableEmailNotificationsValue = ref<boolean>()

    // FIXME: Use graph capability when we have it
    const isSettingsServiceSupported = useCapabilitySpacesEnabled()
    const isChangePasswordEnabled = useCapabilitySpacesEnabled()
    const isPersonalDataExportEnabled = useCapabilityGraphPersonalDataExport()
    const user = computed(() => {
      return store.getters.user
    })

    const showGdprExport = computed(() => {
      return (
        unref(isPersonalDataExportEnabled) &&
        store.getters['runtime/spaces/spaces'].some((s) => isPersonalSpaceResource(s))
      )
    })

    const loadValuesListTask = useTask(function* () {
      try {
        const {
          data: { values }
        } = yield clientService.httpAuthenticated.post('/api/v0/settings/values-list', {
          account_uuid: 'me'
        })
        valuesList.value = values || []
      } catch (e) {
        console.error(e)
        store.dispatch('showMessage', {
          title: $gettext('Unable to load account data…'),
          status: 'danger'
        })
        valuesList.value = []
      }
    }).restartable()

    const loadAccountBundleTask = useTask(function* () {
      try {
        const {
          data: { bundles }
        } = yield clientService.httpAuthenticated.post('/api/v0/settings/bundles-list', {})
        accountBundle.value = bundles?.find((b) => b.extension === 'ocis-accounts')
      } catch (e) {
        console.error(e)
        store.dispatch('showMessage', {
          title: $gettext('Unable to load account data…'),
          status: 'danger'
        })
        accountBundle.value = undefined
      }
    }).restartable()

    const isLoading = computed(() => {
      if (!unref(isSettingsServiceSupported)) {
        return false
      }
      return (
        loadValuesListTask.isRunning ||
        !loadValuesListTask.last ||
        loadAccountBundleTask.isRunning ||
        !loadAccountBundleTask.last
      )
    })

    const languageOptions = computed(() => {
      const languageOptions = unref(accountBundle)?.settings?.find((s) => s.name === 'language')
        ?.singleChoiceValue.options
      return languageOptions?.map((l) => ({
        label: l.displayValue,
        value: l.value.stringValue
      }))
    })

    const groupNames = computed(() => {
      if (unref(useCapabilitySpacesEnabled())) {
        return unref(user)
          .groups.map((group) => group.displayName)
          .join(', ')
      }

      return unref(user).groups.join(', ')
    })

    const saveValue = async ({
      identifier,
      valueOptions
    }: {
      identifier: string
      valueOptions: Record<string, any>
    }) => {
      const valueId = unref(valuesList)?.find((cV) => cV.identifier.setting === identifier)?.value
        ?.id

      const value = {
        bundleId: unref(accountBundle)?.id,
        settingId: unref(accountBundle)?.settings?.find((s) => s.name === identifier)?.id,
        resource: { type: 'TYPE_USER' },
        accountUuid: 'me',
        ...valueOptions,
        ...(valueId && { id: valueId })
      }

      try {
        await clientService.httpAuthenticated.post('/api/v0/settings/values-save', {
          value: {
            accountUuid: 'me',
            ...value
          }
        })

        /**
         * Edge case: we need to reload the values list to retrieve the valueId if not set,
         * otherwise the backend saves multiple entries
         */
        if (!valueId) {
          loadValuesListTask.perform()
        }

        return value
      } catch (e) {
        throw e
      }
    }

    const updateSelectedLanguage = async (option: LanguageOption) => {
      try {
        const value = await saveValue({
          identifier: 'language',
          valueOptions: { listValue: { values: [{ stringValue: option.value }] } }
        })
        selectedLanguageValue.value = option
        setCurrentLanguage({
          language,
          languageSetting: {
            identifier: {
              extension: 'ocis-accounts',
              bundle: 'profile',
              setting: 'language'
            },
            value
          }
        })
        store.dispatch('showMessage', {
          title: $gettext('Language was saved successfully.')
        })
      } catch (e) {
        console.error(e)
        store.dispatch('showMessage', {
          title: $gettext('Saving language failed…'),
          status: 'danger'
        })
      }
    }

    const updateDisableEmailNotifications = async (option: boolean) => {
      try {
        await saveValue({
          identifier: 'disable-email-notifications',
          valueOptions: { boolValue: !option }
        })
        disableEmailNotificationsValue.value = option
        store.dispatch('showMessage', {
          title: $gettext('Email notifications preference saved successfully.')
        })
      } catch (e) {
        console.error(e)
        store.dispatch('showMessage', {
          title: $gettext('Unable to save email notifications preference…'),
          status: 'danger'
        })
      }
    }

    onMounted(async () => {
      if (unref(isSettingsServiceSupported)) {
        await loadAccountBundleTask.perform()
        await loadValuesListTask.perform()

        const languageConfiguration = unref(valuesList)?.find(
          (cV) => cV.identifier.setting === 'language'
        )
        selectedLanguageValue.value = languageConfiguration
          ? unref(languageOptions)?.find(
              (lO) => lO.value === languageConfiguration.value?.listValue?.values?.[0]?.stringValue
            )
          : unref(languageOptions)?.find((o) => o.value === language.current)

        const disableEmailNotificationsConfiguration = unref(valuesList)?.find(
          (cV) => cV.identifier.setting === 'disable-email-notifications'
        )

        disableEmailNotificationsValue.value = disableEmailNotificationsConfiguration
          ? !disableEmailNotificationsConfiguration.value?.boolValue
          : true
      }
    })

    return {
      clientService,
      languageOptions,
      selectedLanguageValue,
      updateSelectedLanguage,
      updateDisableEmailNotifications,
      accountEditLink: store.getters.configuration?.options?.accountEditLink,
      isChangePasswordEnabled,
      showGdprExport,
      isSettingsServiceSupported,
      groupNames,
      user,
      logoutUrl: configurationManager.logoutUrl,
      isLoading,
      disableEmailNotificationsValue,
      loadAccountBundleTask,
      loadValuesListTask
    }
  },
  data() {
    return {
      editPasswordModalOpen: false
    }
  },
  computed: {
    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    showEditPasswordModal() {
      this.editPasswordModalOpen = true
    },
    closeEditPasswordModal() {
      this.editPasswordModalOpen = false
    },
    editPassword(currentPassword, newPassword) {
      return this.clientService.graphAuthenticated.users
        .changeOwnPassword(currentPassword.trim(), newPassword.trim())
        .then(() => {
          this.closeEditPasswordModal()
          this.showMessage({
            title: this.$gettext('Password was changed successfully')
          })
        })
        .catch((error) => {
          console.error(error)
          this.showMessage({
            title: this.$gettext('Failed to change password'),
            status: 'danger'
          })
        })
    }
  }
})
</script>
<style lang="scss">
.account-page-info dd {
  margin-left: 0;
}
</style>
