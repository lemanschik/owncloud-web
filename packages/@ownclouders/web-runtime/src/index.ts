import { DesignSystem as designSystem, pages, translations, supportedLanguages } from './defaults'
import { router } from './router'
import { configurationManager } from 'web-pkg/src/configuration'
import { createHead } from '@vueuse/head'
import { abilitiesPlugin } from '@casl/vue'
import { createMongoAbility } from '@casl/ability'
import merge from 'lodash-es/merge'

import {
  announceConfiguration,
  initializeApplications,
  announceApplicationsReady,
  announceClient,
  announceDefaults,
  announceClientService,
  announceStore,
  announceTheme,
  announceCustomStyles,
  announceTranslations,
  announceVersions,
  announceUppyService,
  announceAuthService,
  startSentry,
  announceCustomScripts,
  announceLoadingService,
  announcePreviewService
} from './container/bootstrap'
import { applicationStore } from './container/store'
import {
  buildPublicSpaceResource,
  buildSpace,
  isPersonalSpaceResource,
  isPublicSpaceResource,
  Resource
} from 'web-client/src/helpers'
import { loadCustomTranslations } from 'web-runtime/src/helpers/customTranslations'
import { WebDAV } from 'web-client/src/webdav'
import { DavProperty } from 'web-client/src/webdav/constants'
import { createApp } from 'vue'
import PortalVue, { createWormhole } from 'portal-vue'

import Avatar from './components/Avatar.vue'
import focusMixin from './mixins/focusMixin'

export const bootstrapApp = async (configurationPath: string): Promise<void> => {
  const app = createApp(pages.success)

  const runtimeConfiguration = await announceConfiguration(configurationPath)
  startSentry(runtimeConfiguration, app)

  const store = await announceStore({ runtimeConfiguration })
  app.use(abilitiesPlugin, createMongoAbility([]), { useGlobalProperties: true })

  const applicationsPromise = initializeApplications({
    app,
    runtimeConfiguration,
    configurationManager,
    store,
    supportedLanguages,
    router,
    translations
  })
  const customTranslationsPromise = loadCustomTranslations({ configurationManager })
  const themePromise = announceTheme({ store, app, designSystem, runtimeConfiguration })
  const [customTranslations] = await Promise.all([
    customTranslationsPromise,
    applicationsPromise,
    themePromise
  ])

  announceTranslations({
    app,
    availableLanguages: supportedLanguages,
    translations: merge(translations, customTranslations)
  })
  announceClientService({ app, runtimeConfiguration, configurationManager, store })
  announceUppyService({ app })
  announceLoadingService({ app })
  announcePreviewService({ app, store, configurationManager })
  await announceClient(runtimeConfiguration)
  announceAuthService({ app, configurationManager, store, router })
  announceCustomStyles({ runtimeConfiguration })
  announceCustomScripts({ runtimeConfiguration })
  announceDefaults({ store, router })

  app.use(router)
  app.use(store)
  app.use(createHead())
  app.config.globalProperties.$wormhole = createWormhole()
  app.use(PortalVue, {
    wormhole: app.config.globalProperties.$wormhole
  })

  app.component('AvatarImage', Avatar)
  app.mixin(focusMixin)

  app.mount('#owncloud')

  const applications = Array.from(applicationStore.values())
  applications.forEach((application) => application.mounted(app))

  store.watch(
    (state, getters) =>
      getters['runtime/auth/isUserContextReady'] ||
      getters['runtime/auth/isIdpContextReady'] ||
      getters['runtime/auth/isPublicLinkContextReady'],
    async (newValue, oldValue) => {
      if (!newValue || newValue === oldValue) {
        return
      }
      announceVersions({ store })
      await announceApplicationsReady({ applications })
    },
    {
      immediate: true
    }
  )

  store.watch(
    (state, getters) => {
      return getters['runtime/auth/isUserContextReady']
    },
    async (userContextReady) => {
      if (!userContextReady) {
        return
      }
      const clientService = app.config.globalProperties.$clientService

      // Load spaces to make them available across the application
      if (store.getters.capabilities?.spaces?.enabled) {
        const graphClient = clientService.graphAuthenticated
        await store.dispatch('runtime/spaces/loadSpaces', { graphClient })
        const personalSpace = store.getters['runtime/spaces/spaces'].find((space) =>
          isPersonalSpaceResource(space)
        )

        if (!personalSpace) {
          return
        }

        store.commit('runtime/spaces/UPDATE_SPACE_FIELD', {
          id: personalSpace.id,
          field: 'name',
          value: app.config.globalProperties.$gettext('Personal')
        })
        return
      }

      // Spaces feature not available. Create a virtual personal space
      const user = store.getters.user
      const space = buildSpace({
        id: user.id,
        driveAlias: `personal/${user.id}`,
        driveType: 'personal',
        name: app.config.globalProperties.$gettext('All files'),
        webDavPath: `/files/${user.id}`,
        webDavTrashPath: `/trash-bin/${user.id}`,
        serverUrl: configurationManager.serverUrl
      })
      const personalHomeInfo = await (clientService.webdav as WebDAV).getFileInfo(
        space,
        {
          path: ''
        },
        { davProperties: [DavProperty.FileId] }
      )
      space.fileId = personalHomeInfo.fileId
      store.commit('runtime/spaces/ADD_SPACES', [space])
      store.commit('runtime/spaces/SET_SPACES_INITIALIZED', true)
    },
    {
      immediate: true
    }
  )
  store.watch(
    (state, getters) => {
      return getters['runtime/auth/isPublicLinkContextReady']
    },
    (publicLinkContextReady) => {
      if (!publicLinkContextReady) {
        return
      }
      // Create virtual space for public link
      const publicLinkToken = store.getters['runtime/auth/publicLinkToken']
      const publicLinkPassword = store.getters['runtime/auth/publicLinkPassword']
      const space = buildPublicSpaceResource({
        id: publicLinkToken,
        name: app.config.globalProperties.$gettext('Public files'),
        ...(publicLinkPassword && { publicLinkPassword }),
        serverUrl: configurationManager.serverUrl
      })
      store.commit('runtime/spaces/ADD_SPACES', [space])
      store.commit('runtime/spaces/SET_SPACES_INITIALIZED', true)
    },
    {
      immediate: true
    }
  )
  store.watch(
    // only needed if a public link gets re-resolved with a changed password prop (changed or removed).
    // don't need to set { immediate: true } on the watcher.
    (state, getters) => {
      return getters['runtime/auth/publicLinkPassword']
    },
    (publicLinkPassword: string | undefined) => {
      const publicLinkToken = store.getters['runtime/auth/publicLinkToken']
      const space = store.getters['runtime/spaces/spaces'].find((space: Resource) => {
        return isPublicSpaceResource(space) && space.id === publicLinkToken
      })
      if (!space) {
        return
      }
      space.publicLinkPassword = publicLinkPassword
    }
  )
}

export const bootstrapErrorApp = async (err: Error): Promise<void> => {
  const store = await announceStore({ runtimeConfiguration: {} })
  announceVersions({ store })
  const app = createApp(pages.failure)
  await announceTheme({ store, app, designSystem })
  console.error(err)
  app.use(store)
  await announceTranslations({
    app,
    availableLanguages: supportedLanguages,
    translations
  })
  app.mount('#owncloud')
}
;(window as any).runtimeLoaded({
  bootstrapApp,
  bootstrapErrorApp
})
