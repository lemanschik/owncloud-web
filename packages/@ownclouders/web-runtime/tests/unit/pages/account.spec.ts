import account from '../../../src/pages/account.vue'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  mockAxiosResolve,
  shallowMount,
  defaultStoreMockOptions,
  mockAxiosReject
} from 'web-test-helpers'
import { mock, mockDeep } from 'jest-mock-extended'
import { SpaceResource } from 'web-client/src/helpers'
import { AxiosResponse } from 'axios'
import { ClientService } from 'web-pkg'
import { SettingsBundle, SettingsValue } from 'web-runtime/src/helpers/settings'

const $route = {
  meta: {
    title: 'Some Title'
  }
}

const selectors = {
  pageTitle: '.oc-page-title',
  loaderStub: 'oc-spinner-stub',
  editUrlButton: '[data-testid="account-page-edit-url-btn"]',
  accountPageInfo: '.account-page-info',
  groupNames: '[data-testid="group-names"]',
  groupNamesEmpty: '[data-testid="group-names-empty"]',
  gdprExport: '[data-testid="gdpr-export"]'
}

jest.mock('web-pkg/src/configuration', () => ({
  configurationManager: {
    logoutUrl: 'https://account-manager/logout'
  }
}))

describe('account page', () => {
  describe('header section', () => {
    it('renders page title', () => {
      const { wrapper } = getWrapper()
      const pageTitle = wrapper.find(selectors.pageTitle)
      expect(pageTitle.exists()).toBeTruthy()
      expect(pageTitle.text()).toBe($route.meta.title)
    })

    describe('edit url button', () => {
      it('should be displayed if defined via config', () => {
        const { wrapper } = getWrapper({
          accountEditLink: { href: '/' },
          data: { loadingGroups: false }
        })
        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.html()).toMatchSnapshot()
      })
      it('should not be displayed if not defined via config', () => {
        const { wrapper } = getWrapper({
          data: { loadingGroups: false }
        })
        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.exists()).toBeFalsy()
      })
    })
  })
  describe('account information', () => {
    it('displays basic user information', () => {
      const { wrapper } = getWrapper({
        user: {
          user: {
            username: 'some-username',
            displayname: 'some-displayname',
            email: 'some-email'
          }
        }
      })

      const accountPageInfo = wrapper.find(selectors.accountPageInfo)
      expect(accountPageInfo.html()).toMatchSnapshot()
    })

    describe('group membership', () => {
      it('displays message if not member of any groups', () => {
        const { wrapper } = getWrapper({ user: { groups: [] } })

        const groupNamesEmpty = wrapper.find(selectors.groupNamesEmpty)
        expect(groupNamesEmpty.exists()).toBeTruthy()
      })
      it('displays group names', () => {
        const { wrapper } = getWrapper({ user: { groups: ['one', 'two', 'three'] } })

        const groupNames = wrapper.find(selectors.groupNames)
        expect(groupNames.html()).toMatchSnapshot()
      })
    })
  })

  describe('gdpr export section', () => {
    it('does show if announced via capabilities and user has a personal space', () => {
      const spaces = [mock<SpaceResource>({ driveType: 'personal' })]
      const { wrapper } = getWrapper({
        spaces,
        capabilities: { graph: { 'personal-data-export': true } }
      })
      expect(wrapper.find(selectors.gdprExport).exists()).toBeTruthy()
    })
    it('does not show if not announced via capabilities', () => {
      const spaces = [mock<SpaceResource>({ driveType: 'personal' })]
      const { wrapper } = getWrapper({
        spaces,
        capabilities: { graph: { 'personal-data-export': false } }
      })
      expect(wrapper.find(selectors.gdprExport).exists()).toBeFalsy()
    })
    it('does not show if user has no personal space', () => {
      const spaces = [mock<SpaceResource>({ driveType: 'project' })]
      const { wrapper } = getWrapper({
        spaces,
        capabilities: { graph: { 'personal-data-export': true } }
      })
      expect(wrapper.find(selectors.gdprExport).exists()).toBeFalsy()
    })
  })

  describe('method "editPassword"', () => {
    it('should show message on success', async () => {
      const { wrapper, mocks } = getWrapper()
      mocks.$clientService.graphAuthenticated.users.changeOwnPassword.mockResolvedValue(
        mockAxiosResolve()
      )
      const showMessageStub = jest.spyOn(wrapper.vm, 'showMessage')
      await wrapper.vm.editPassword('password', 'newPassword')
      expect(showMessageStub).toHaveBeenCalled()
    })

    it('should show message on error', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper, mocks } = getWrapper()
      mocks.$clientService.graphAuthenticated.users.changeOwnPassword.mockRejectedValue(new Error())
      const showMessageStub = jest.spyOn(wrapper.vm, 'showMessage')
      await wrapper.vm.editPassword('password', 'newPassword')
      expect(showMessageStub).toHaveBeenCalled()
    })
  })

  describe('computed method "isChangePasswordEnabled"', () => {
    it('should be true if capability is enabled', () => {
      const { wrapper } = getWrapper({ capabilities: { spaces: { enabled: true } } })
      expect(wrapper.vm.isChangePasswordEnabled).toBeTruthy()
    })
    it('should be false if capability is not enabled', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.vm.isChangePasswordEnabled).toBeFalsy()
    })
  })
  describe('Logout from all devices link', () => {
    it('should render the logout from active devices if logoutUrl is provided', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
    })
    it("shouldn't render the logout from active devices if logoutUrl isn't provided", () => {
      const { wrapper } = getWrapper()
      wrapper.vm.logoutUrl = undefined
      expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
    })
    it('should use url from configuration manager', () => {
      const { wrapper } = getWrapper()
      const button = wrapper.find('oc-button-stub')
      expect(button.attributes('href')).toBe('https://account-manager/logout')
    })
  })
  describe('Method "updateDisableEmailNotifications', () => {
    it('should show a message on success', async () => {
      const clientServiceMock = mockDeep<ClientService>()
      clientServiceMock.httpAuthenticated.post.mockResolvedValueOnce(
        mock<AxiosResponse>({ data: { bundles: [mock<SettingsBundle>()] } })
      )
      clientServiceMock.httpAuthenticated.post.mockResolvedValueOnce(
        mock<AxiosResponse>({ data: { values: [mock<SettingsValue>()] } })
      )

      const { wrapper, mocks, storeOptions } = getWrapper({
        clientServiceMock,
        capabilities: { spaces: { enabled: true } }
      })

      await wrapper.vm.loadAccountBundleTask.last
      await wrapper.vm.loadValuesListTask.last

      mocks.$clientService.httpAuthenticated.post.mockResolvedValueOnce(mockAxiosResolve({}))
      await wrapper.vm.updateDisableEmailNotifications(true)
      expect(storeOptions.actions.showMessage).toHaveBeenCalledWith(
        expect.anything(),
        expect.not.objectContaining({ status: 'danger' })
      )
    })
    it('should show a message on error', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)

      const clientServiceMock = mockDeep<ClientService>()
      clientServiceMock.httpAuthenticated.post.mockResolvedValueOnce(
        mock<AxiosResponse>({ data: { bundles: [mock<SettingsBundle>()] } })
      )
      clientServiceMock.httpAuthenticated.post.mockResolvedValueOnce(
        mock<AxiosResponse>({ data: { values: [mock<SettingsValue>()] } })
      )

      const { wrapper, mocks, storeOptions } = getWrapper({
        clientServiceMock,
        capabilities: { spaces: { enabled: true } }
      })

      await wrapper.vm.loadAccountBundleTask.last
      await wrapper.vm.loadValuesListTask.last

      mocks.$clientService.httpAuthenticated.post.mockImplementation(() => mockAxiosReject('err'))
      await wrapper.vm.updateDisableEmailNotifications(true)
      expect(storeOptions.actions.showMessage).toHaveBeenCalledWith(
        expect.anything(),
        expect.objectContaining({ status: 'danger' })
      )
    })
  })
})

function getWrapper({
  data = {},
  user = {},
  capabilities = {},
  accountEditLink = undefined,
  spaces = [],
  clientServiceMock = mockDeep<ClientService>()
} = {}) {
  const storeOptions = { ...defaultStoreMockOptions }
  storeOptions.modules.runtime.modules.spaces.getters.spaces.mockReturnValue(spaces)
  storeOptions.getters.user.mockReturnValue({ groups: [], ...user })
  storeOptions.getters.capabilities.mockReturnValue(capabilities)
  storeOptions.getters.configuration.mockReturnValue({
    server: 'http://server/address/',
    options: { ...(accountEditLink && { accountEditLink }) }
  })

  const store = createStore(storeOptions)
  const mocks = {
    ...defaultComponentMocks(),
    $route
  }
  mocks.$clientService = clientServiceMock
  return {
    storeOptions,
    mocks,
    wrapper: shallowMount(account, {
      data: () => data,
      global: {
        plugins: [...defaultPlugins(), store],
        mocks,
        stubs: {
          'oc-button': true,
          'oc-icon': true
        }
      }
    })
  }
}
