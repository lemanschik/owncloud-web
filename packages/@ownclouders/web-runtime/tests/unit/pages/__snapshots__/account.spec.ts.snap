// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account page account information displays basic user information 1`] = `
<dl class="account-page-info oc-flex oc-flex-wrap">
  <div class="account-page-info-username oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Username</dt>
    <dd></dd>
  </div>
  <!--v-if-->
  <div class="account-page-info-displayname oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Display name</dt>
    <dd></dd>
  </div>
  <div class="account-page-info-email oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Email</dt>
    <dd>
      <span>No email has been set up</span>
    </dd>
  </div>
  <div class="account-page-info-groups oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Group memberships</dt>
    <dd>
      <span>You are not part of any group</span>
    </dd>
  </div>
  <!--v-if-->
  <div class="account-page-logout-all-devices oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Logout from active devices</dt>
    <dd>
      <oc-button-stub appearance="raw" href="https://account-manager/logout" target="_blank" type="a"></oc-button-stub>
    </dd>
  </div>
  <!--v-if-->
  <!--v-if-->
</dl>
`;

exports[`account page account information group membership displays group names 1`] = `
<dd>
  <span>one, two, three</span>
</dd>
`;

exports[`account page header section edit url button should be displayed if defined via config 1`] = `<oc-button-stub href="/" target="_blank" type="a" variation="primary"></oc-button-stub>`;
